### PR TITLE
Relax broadcasting constraint on scale_tensor

### PR DESCRIPTION
**Pull request by fritzo**
This is needed for parallel enumeration, where e.g. the a sample site's values are the same in all enumeration branches (leading to a log_prob that is not broadcast), but a `poutine.scale` mask does depend on the enumeration values. In this case we need to broadcast up the site log_prob to handle the `scale` of larger shape.

## Tested

- added exhaustive tests
- tested on an internal application

### CODE REVIEW

File: pyro/distributions/util.py [hunk 1]
```diff
@@ -148,30 +148,34 @@ def sum_leftmost(value, dim):
...
-        else:
-            return tensor * scale
+
     result = tensor * scale
-    result[(scale == 0).expand_as(result)] = 0  # avoid NANs
-    if result.shape != tensor.shape:
-        raise ValueError("Broadcasting error: scale is incompatible with tensor: "
-                         "{} vs {}".format(scale.shape, tensor.shape))
+
+    # avoid NANs
```
Comment (martinjankowiak) [inline pyro/distributions/util.py]: why do you want these?

Comment (fritzo) [reply pyro/distributions/util.py]: The main reason is to allow `poutine.scale` to completely omit some senseless log prob terms by enforcing hard constraints.

Comment (fritzo) [reply pyro/distributions/util.py]: The main reason is to allow `poutine.scale` to completely omit some senseless log prob terms by enforcing hard constraints.

Comment (jpchen) [reply pyro/distributions/util.py]: to safely mask out any zero scale values when `poutine.scale`ing

**Pull request merged**